add LocalStorageKey; rework LocalStorage API

lukaskollmer opened the pull request:
# add LocalStorageKey; rework LocalStorage API

## :recycle: Current situation & Problem
`LocalStorage` currently identifies entries in the storage via either optional, raw untyped strings, or via the static type of the to-be-stored/loaded value.
While this approach does lead to a nice-looking API, it is problematic for several reasons:
- identifying stored entries using string literals makes it easy to misspel something, or to accidentally use the wrong key
- the fact that there's no static coupling between a key and the type it's used to store/load, means that there's nothing that would prevent someone from acidentally decoding a value into an incorrect type (i.e., a different type than what was encoded into the `LocalStorage` entry)
- the fact that parameters such as e.g. the encoder/decoder or the storage setting, both of which determine how a value is stored/loaded, need to be explicitly specified at both calls to `store()` as well as calls to `read()`, means that it is easy to mismatch these parameters
  - eg: if you use a `PropertyListEncoder` when storing some value into the `LocalStorage`, you need to make sure that you also use the matching `PropertyListDecoder` when reading the stored value.
    - since the current API defaults both the encoder in `store()` and the decoder in `read()` to the `JSONEncoder` (and `JSONDecoder, respectively), makes it too easy to miss the parameter somewhere
  - the same applies to the `storageSetting` parameter: all matching calls to `store()` and `read()` that operate on the same storage key must specify the same storage setting.
    - this is currently not enforced, and with the current API there is no way of enforcing this
- the fact that storage keys are optional (if the key is nil (the default), the unqualified typename of the to-be-stored/loaded type will be used instead) is also likely to lead to unexpected issues:
  - since the unqualified typename doesn't take nesting / access control into account, it can easily happen that two calls to `store()`/`read()` that operate on completely different types (potentially even in completely different modules!) end up operating on the same LocalStorage entry, because the unqualified typename will be the same in both cases

This PR attempts to address these issues, by restructuring and redesigning the `LocalStorage` API:
- Entries in the storage are no longer identified by raw strings which are passed to the `store()`/`load()` calls
- Entries are no longer identified by the typename of the value being stored/loaded
- We instead have a `LocalStorageKey<Value>` type, which are intended to be long-lived objects that are used to access the `LocalStorage`
  - the `LocalStorageKey` encapsulates all information required to store data into the storage / load data from it
  - since every store/load for an entry will use the same `LocalStorageKey` instance, we can ensure that every store/load for the same entry will use matching decoders/encoders, and will use the same storageSetting.
  - it also improves the typing, by associating every storage key with its underlying stored type
  - in order to prevent race conditions and other potential concurrency issues, the `LocalStorageKey` also acts as a read-write lock when accessing the storage entry identified by it
- In addition to these changes, the `LocalStorageKey`-based API also makes it easier to support storing additional currently not supported types, such as e.g. `NSSecureCoding`-compliant types
- **Note** this PR removes support for `CodableWithConfiguration`, which previously was supported but (AFAICT) not used.
- **Note**: `SpeziSecureStorage` is currently untouched; it might make sense to have something similar there as well, but it'd of course require a separate, dedicated key type and would be out-of-scope for this PR.
- This PR also adds a small convenience property wrapper, `@LocalStorageEntry`, which makes it easier to use the `LocalStorage` API from within a SwiftUI View, and is e.g. able to observe a key's value and update the view in response.


## :gear: Release Notes 
- Completely reworked the `LocalStorage` API:
  - Values are stored/loaded via `LocalStorageKey`s, which define an entry in the storage.
  - `LocalStorageKey`s need to explicitly specify their underlying raw keys (a String), and are encouraged to use reverse-DNS-notation in order to avoid collisions


## :books: Documentation
The documentation has been adapted to the new API.

## :white_check_mark: Testing
The tests have been adapted to the new API.


## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

lukaskollmer committed:
add LocalStorageKey; rework LocalStorage API

lukaskollmer committed:
make it pretty

lukaskollmer committed:
hmmm

lukaskollmer committed:
add @LocalStorageEntry property wrapper

lukaskollmer committed:
add tests for @LocalStorageEntry

lukaskollmer committed:
make swiftlint happy, learn how combine works

(the second part is still work in progress)

lukaskollmer committed:
mmmmm

lukaskollmer committed:
some > any

lukaskollmer commented on Sources/SpeziLocalStorage/LocalStorageEntry.swift:
i'm not overly experienced when it comes to using Combine, so please let me know whether what i'm doing here makes sense / is the correct approach

philippzagar replied:
Looks alright to me, even tho I'm also rather inexperienced with Combine.

PSchmiedmayer replied:
Looks good to me; only had once cancellation & cleanup comment 👍

philippzagar commented on Tests/UITests/TestApp/LocalStorageTests/LocalStorageLiveUpdateTestView.swift:
```diff
@@ -0,0 +1,43 @@
+}
+
+
+struct LocalStorageLiveUpdateTestView: View { // swiftlint:disable:this file_types_order
```
Cool test, thanks!

philippzagar commented on Tests/UITests/TestAppUITests/LocalStorageTests.swift:
```diff
@@ -14,9 +14,22 @@ final class LocalStorageTests: XCTestCase {
+        let numbers = (0..<17).map { _ in Int.random(in: 0..<5) }
+        for number in numbers {
+            app.buttons["\(number)"].tap()
+            XCTAssertTrue(app.staticTexts.matching(NSPredicate(format: "label MATCHES %@", "number.*\(number)")).element.waitForExistence(timeout: 0.5))
```
The timeouts might be a bit low for the CI (we usually have around 2 sec), but we can evolve that if needed in the future.

lukaskollmer replied:
i added the small delay since not having any delay would in some cases not catch the update (afaik the update happens on the next run loop invocation, so there is a small delay we need to account for)

philippzagar commented on Sources/SpeziLocalStorage/SpeziLocalStorage.docc/SpeziLocalStorage.md:
````diff
@@ -64,10 +64,23 @@ Alternatively, it is common to use the `LocalStorage` module in other modules as
 
 
-### Storing & Update Data
+### Defining Storage Keys
+
+`LocalStorage` uses unique ``LocalStorageKey``s to .
+
+You define storage keys by placing a static non-computed properties of type ``LocalStorageKey`` into an extension on the ``LocalStorageKeys`` type:
+
+```swift
+extension LocalStorageKeys {
+    static let note = LocalStorageKey
+}
+```
+
+
+### Storing and Loading Data
````
Thanks for the updated docs! 🚀  I feel there are quite some other doc parts that we need to updated in the README and DocC

lukaskollmer replied:
good catch; i'll look over the docs and make sure everything is in sync and up to date

philippzagar commented on Sources/SpeziLocalStorage/LocalStorageKey.swift:
```diff
@@ -0,0 +1,145 @@
+    public init(
+        key: String,
+        setting: LocalStorageSetting = .default,
+        encode: @Sendable @escaping (Value) throws -> Data,
+        decode: @Sendable @escaping (Data) throws -> Value?
```
Cool idea, I like it!

philippzagar commented on Sources/SpeziLocalStorage/LocalStorageKey.swift:
```diff
@@ -0,0 +1,145 @@
+    let setting: LocalStorageSetting
+    let encode: @Sendable (Value) throws -> Data
+    let decode: @Sendable (Data) throws -> Value?
+    private let lock = RWLock()
```
Just as a long shot: Should we aim to move away from manual locking and go towards an async/await actor-based implementation here? This would of course then also propagate to the public layer, which probably is not wanted. Might be overkill, just putting it out there.

lukaskollmer replied:
that was @PSchmiedmayer's idea as well, but there's, from what i can tell, some issues with that approach:
- with actors, we wouldn't be able to allow concurrent reads while enforcing exclusive writes (which currently is allowed).
  whether this is something we actually need is a different question of course (the answer is almost definitely no, but i'd argue it does feel more elegant this way)
- were we to use actors, we'd need to make all store() / load() calls async (that'd be the entire point, of course), which might be overkill, since the likelihood of parallel reads is already pretty low (plus, with the current implementation parallel reads would be fine anyway), and parallel writes are already synchronized currently.
- more importantly, we wouldn't be able to have a property-wrapper for accessing storage entries anymore, since the `wrappedValue` can't be async.

philippzagar replied:
Yep I share these concerns as well, I think the downsides of actor isolation outway the benefits in this case, so from my side I'm fine with keeping the `RWLock`, even tho it doesn't feel super swifty.

PSchmiedmayer replied:
Agree with @lukaskollmer here; we discussed this a bit and I think that the basic locking mechanism should be fine for these operations and they should be atomic anyways. The fact that we can do concurrent reads is a big plus and most likely the biggest use case. Ideally larger file store operations are conducted in an asynchronous background task anyways and we might expand our documentation to hint to that?

philippzagar replied:
I agree with everything said above, in this case the simple `RWLock` provides nice properties that an actor mechanism can't.

philippzagar commented on Sources/SpeziLocalStorage/LocalStorageEntry.swift:
```diff
@@ -0,0 +1,49 @@
+
+
+/// Access ``LocalStorage`` entries within a SwiftUI View.
+@propertyWrapper
```
Great idea to introduce that! 🚀 
Would be great to get a small usage doc in here for the wrapper as well as writing down the need for the `LocalStorage` module in the env.

philippzagar replied:
@lukaskollmer Might make sense to define the wrapper like this, just on a `View`?
https://github.com/StanfordSpezi/SpeziLLM/blob/26b1e07756556b1b329f3a9a3267a5f99c0c2e78/Sources/SpeziLLMOpenAI/FunctionCalling/LLMFunctionParameterWrapper.swift#L83

lukaskollmer replied:
not sure what you mean / how that would would work. `@Property` is a property wrapper which gets its value externally injected (as part of handling the LLM function call), which isn't something we can do with SwiftUI (if that's what you're talking about).

or are you referring to some other aspect of `@Parameter`?

philippzagar replied:
I mean defining the property wrapper to only be directly accessible within a `View` (so without a `View.ABC` prefix), so something like:
```swift
extension View {
    typealias ABC = _PropertyWrapper...
}
```

However, I think SwiftUI isn't doing something like that for its property wrappers as well, so it was just a random thought I had to constrain the wrapper usage.

lukaskollmer replied:
oh ok. i actually really like this pattern but i agree that it probably should not happen in this PR specifically (we probably should also adopt this in other packages as well were we to adopt it here...)

philippzagar commented on Sources/SpeziLocalStorage/LocalStorage.swift:
```diff
@@ -245,64 +169,85 @@ public final class LocalStorage: Module, DefaultInitializable, EnvironmentAccess
+    ///     If the closure returns `nil`, the entry will be removed from the `LocalStorage`.
+    ///
+    /// - throws: if `transform` throws,
+    public func modify<Value>(_ key: LocalStorageKey<Value>, transform: (Value?) throws -> Value?) throws {
```
Nit: Would be great to test this function

philippzagar commented on Sources/SpeziLocalStorage/LocalStorage.swift:
```diff
@@ -70,74 +68,36 @@ public final class LocalStorage: Module, DefaultInitializable, EnvironmentAccess
+            if let value {
+                try storeImp(value, for: key)
+            } else {
+                try delete(key)
```
Nit: Would be great to test this branch

philippzagar commented on Tests/UITests/TestApp/LocalStorageTests/LocalStorageLiveUpdateTestView.swift:
```diff
@@ -0,0 +1,43 @@
+    @LocalStorageEntry(.number) private var number
+    
+    var body: some View {
+        LabeledContent("number", value: number.map(String.init) ?? "–")
```
Maybe test a small write to the property wrapper as well?

philippzagar reviewed (approved):
Thanks for the great PR @lukaskollmer! 🚀 
Mostly had some smaller nits and long-shot ideas. What's important is to update the rest of the documentation.
And yes, a similar PR in the SecureStorage would be great as well!

lukaskollmer committed:
also test @LocalStorageEntry setter

lukaskollmer committed:
change `modify` signature, add a test case

lukaskollmer committed:
swiftlint

lukaskollmer committed:
extend test coverage, fix a bug along the way

who would've thought testing the code would reveal the bugs?

PSchmiedmayer commented on Sources/SpeziLocalStorage/LocalStorage.swift:
```diff
@@ -171,69 +130,42 @@ public final class LocalStorage: Module, DefaultInitializable, EnvironmentAccess
+    
+    
+    /// Determines whether the `LocalStorage` contains a value for the specified key.
+    public func hasEntry(for key: LocalStorageKey<some Any>) -> Bool {
```
Nice addition! 🚀

PSchmiedmayer commented on Sources/SpeziLocalStorage/LocalStorage.swift:
```diff
@@ -245,64 +177,85 @@ public final class LocalStorage: Module, DefaultInitializable, EnvironmentAccess
-    ///   - type: The `Encodable` type that is used to store the type originally.
-    public func delete<C: Encodable>(_ type: C.Type = C.self) throws {
-        try delete(C.self, storageKey: nil)
+    ///   - key: An optional storage key to identify the file.
```
Actually might not be "optional" is it?

PSchmiedmayer commented on Sources/SpeziLocalStorage/LocalStorage.swift:
```diff
@@ -245,64 +177,85 @@ public final class LocalStorage: Module, DefaultInitializable, EnvironmentAccess
+    ///     If the closure sets `value` to `nil`, the entry will be removed from the `LocalStorage`.
+    ///
+    /// - throws: if `transform` throws,
+    public func modify<Value>(_ key: LocalStorageKey<Value>, transform: (_ value: inout Value?) throws -> Void) throws {
```
Also a nice addition 🚀

PSchmiedmayer commented on Sources/SpeziLocalStorage/LocalStorageEntry.swift:
```diff
@@ -0,0 +1,66 @@
+private final class LocalStorageEntryInternals<Value> {
+    fileprivate var value: Value?
+    
+    @ObservationIgnored private var cancellable: AnyCancellable?
```
We should probably cancel the observation on deinit of the class here? Would then be triggered if `LocalStorageEntry` is no longer referenced in a view?

lukaskollmer replied:
the docs say that the `AnyCancellable` will auto-cancel itself when it gets deinitialized

PSchmiedmayer replied:
Sounds good 👍

PSchmiedmayer commented on Sources/SpeziLocalStorage/LocalStorageKey.swift:
```diff
@@ -0,0 +1,145 @@
+/// }
+///
+/// extension LocalStorageKeys {
+///     static let note = LocalStorageKey<Note>("note")
+///     static let person = LocalStorageKey<Person>("person", encoder: PropertyListEncoder(), decoder: PropertyListDecoder())
```
We should probably follow our documentation and use reverse-DNS-keys here 👍

PSchmiedmayer commented on Sources/SpeziLocalStorage/LocalStorageKey.swift:
```diff
@@ -0,0 +1,145 @@
+/// Defining and using two or more `LocalStorageKey`s with identical underlying `key`s will lead to conflicts and data loss or corruption.
+/// It is advised that you use a reverse-DNS-style naming scheme for your keys, in order to avoid name collisions.
+///
+/// ### Supported Types
```
Nice additions!

PSchmiedmayer reviewed (approved):
Very cool additions, I agree that we should apply the same mechanism to the Secure Storage component as well. 

Great additional tests and coverage; only had a few minor comments.

lukaskollmer committed:
improve docs

lukaskollmer committed:
update docs

lukaskollmer committed:
fix access control change

lukaskollmer committed:
update readme

i hope the urls are correct; i'm mostly just guessing here based on what i assume it'll get translated into

lukaskollmer committed:
update docs

lukaskollmer committed:
add a test case for raw `Data` storage

lukaskollmer commented:
@PSchmiedmayer i agree that we should introduce a similar API for the `SecureStorage` (and maybe also do the target renaming into `SpeziStorage` and `SpeziCredentialsStorage`, or `SpeziStorage` and `SpeziSecureStorage`), but IMO that should be a separate PR.

lukaskollmer committed:
add a test case for the NSSecureStorage path

PSchmiedmayer commented:
Sounds great; happy to see the PR merged & we can create a GitHub issue for the Secure Storage mechanism & ideally get that in place before we tag a 2.0.
